[Learn] Abstraction and Encapsulation pathway (#875)
pulumi/pulumi-hugo: themes/default/content/learn/abstraction-encapsulation/encapsulation/index.md

---
title: "Understanding Encapsulation"
layout: topic
date: 2021-11-17
draft: false
description: |
    Explore encapsulation of cloud resources with Pulumi.
meta_desc: |
    Explore encapsulation of cloud resources with Pulumi.
index: 0
estimated_time: 10

> meta_image: meta.png

Reviewer: Lil' reminder to update this.

Author: ✅  done

authors:
    - laura-santamaria
tags:
    - learn
block_external_search_index: false
---

So we've explored a bit about abstraction. What about encapsulation?

Encapsulation is the logical conclusion of an abstracted system: Breaking logic
down into reusable components. In doing so, we ensure better maintainability,
readability, and reusability. To illustrate that, let's explore a practical
example.

## Breaking apart Pulumi code

Let's take a fairly uncomplicated piece of Pulumi code: the definition of an s3
bucket:

```python
import pulumi

> import pulumi_aws_native as aws

Reviewer 2: Mmmmmmmaybe import `as aws_native` to make it glaringly obvious which provider is being used?

Reviewer 2: Also consider just using AWS classic as referenced below to keep everything within a single provider, since AWS Native isn't going to be RTM until 2022.

Author: Good choice on switching it to `aws_native`. I've been just using `aws` in workshops and such, but I'm there to keep reminding the attendee of the import naming choice. Fixing in the next commit.

In regard to the question around the provider, we're trying to using the native provider whenever possible so folks get familiar with it. While I can see the argument, I think I'll stick with it as we have it. This part of the API is pretty stable as it's the one they're using to show off everything they're doing.

bucket = aws.s3.bucket("my-bucket")

pulumi.export("bucket", bucket.bucket_name)
```

Here, we're creating one resource and exporting the output from that resource.

> Generally, however, there's a lot more to building up infrastructure with a

Reviewer 2: ```suggestion
In practice however, production systems typically have more requirements than a
```

                    "s3:GetObject"
                ],
                "Resource": [
                    f"{arn}/*"
                ]
            }]
        })
    ),
    opts=pulumi.ResourceOptions(parent=bucket)
)

pulumi.export("bucket", bucket.bucket_name)
```

Now, that's nice, but what if we need to make three different storage objects
for three different use cases? We could start out by creating a resource
grouping for the kind of resource we wanted to use:

```python
import ...

class OurBucketClass(self, name_me):
    bucket = aws.s3.bucket(f"{name_me}")
    bucket_policy = aws_classic.s3.BucketPolicy(
        f"{name_me}-policy",
        bucket=bucket.id,
        policy=bucket.arn.apply(
            lambda arn: json.dumps({
                "Version": "2012-10-17",
                "Statement": [{
                    "Effect": "Allow",
                    "Principal": "*",
                    "Action": [
                        "s3:GetObject"
                    ],
                    "Resource": [
                        f"{arn}/*"
                    ]
                }]
            })
        ),
        opts=pulumi.ResourceOptions(parent=bucket)
    )
```

> But that JSON list of the policy is an object, too, that we can tuck away,

Reviewer 2: ```suggestion
But that JSON blob of the policy is also an object which we can encapsulate,
```

making our storage object class more usable outside of this specific context.
Let's imagine we have a referenceable list of common access policies, perhaps
a simple key:value store. In the store, keys are strings that are names of
policies, and the values are JSON documents.

```python
import ...

class OurBucketClass(self, name_me, policy_name):

    policy_list = {
        'default': '{...}',
        'locked' : '{...}',
        'permissive': '{...}'
    }

    def define_policy(policy_name, bucket_id):

>         try:

Reviewer 2: This `try` should be removed as it has no effect since the `except` block has no error handling.

Author: whoops, forgot that; I'm fixing the block with better handling in the next commit
